shell/browser/net/intercepting_url_loader_factory.cc

```diff
@@ -0,0 +1,189 @@
+}
+
+InterceptingURLLoaderFactory::~InterceptingURLLoaderFactory() = default;
+
+bool InterceptingURLLoaderFactory::ShouldIgnoreConnectionsLimit(
+    const network::ResourceRequest& request) {
+  for (const auto& domain : ignore_connections_limit_domains_) {
+    if (request.url.DomainIs(domain)) {
+      return true;
+    }
+  }
+  return false;
+}
+
+void InterceptingURLLoaderFactory::CreateLoaderAndStart(
+    mojo::PendingReceiver<network::mojom::URLLoader> loader,
+    int32_t routing_id,
+    int32_t request_id,
+    uint32_t options,
+    const network::ResourceRequest& original_request,
+    mojo::PendingRemote<network::mojom::URLLoaderClient> client,
+    const net::MutableNetworkTrafficAnnotationTag& traffic_annotation) {
+  // Take a copy so we can mutate the request.
+  network::ResourceRequest request = original_request;
+
+  if (ShouldIgnoreConnectionsLimit(request)) {
+    request.priority = net::RequestPriority::MAXIMUM_PRIORITY;
+    request.load_flags |= net::LOAD_IGNORE_LIMITS;
+  }
+
+  // Check if user has intercepted this scheme.
+  auto it = intercepted_handlers_.find(request.url.scheme());
+  if (it != intercepted_handlers_.end()) {
+    mojo::PendingRemote<network::mojom::URLLoaderFactory> this_remote;
+    this->Clone(this_remote.InitWithNewPipeAndPassReceiver());
+
+    mojo::PendingRemote<network::mojom::URLLoaderFactory>
+        proxying_factory_remote;
+    proxying_factory_remote_->Clone(
+        proxying_factory_remote.InitWithNewPipeAndPassReceiver());
+
+    scoped_refptr<InterceptedRequest> intercepting_loader =
+        base::MakeRefCounted<InterceptedRequest>(
+            std::move(loader), routing_id, request_id, options, request,
+            std::move(client), traffic_annotation, it->second.first,
+            std::move(this_remote), std::move(proxying_factory_remote));
+
+    // <scheme, <type, handler>>
+    it->second.second.Run(
+        request,
+        base::BindOnce(&InterceptedRequest::SendResponse, intercepting_loader),
+        base::BindOnce(&InterceptedRequest::ContinueRequest,
+                       intercepting_loader));
+  } else {
+    // Not intercepted, continue on to the proxying factory
+    proxying_factory_remote_->CreateLoaderAndStart(
+        std::move(loader), routing_id, request_id, options, request,
+        std::move(client), traffic_annotation);
+  }
+}
+
+void InterceptingURLLoaderFactory::Clone(
+    mojo::PendingReceiver<network::mojom::URLLoaderFactory> loader_receiver) {
+  interceptor_receivers_.Add(this, std::move(loader_receiver));
+}
+
+void InterceptingURLLoaderFactory::OnProxyingFactoryError() {
+  proxying_factory_remote_.reset();
+  interceptor_receivers_.Clear();
+
+  MaybeDeleteThis();
+}
+
+void InterceptingURLLoaderFactory::OnInterceptorBindingError() {
+  if (interceptor_receivers_.empty())
+    proxying_factory_remote_.reset();
+
+  MaybeDeleteThis();
+}
+
+void InterceptingURLLoaderFactory::MaybeDeleteThis() {
```

Comment:
Is this only ever deleted when there's an error...? Or is this called in the normal lifecycle? If the latter, perhaps the methods could be renamed `Error` -> `Disconnect`?